scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java

```diff
@@ -1303,28 +1320,51 @@ protected final void expectChangeCipherSpecMessage() {
 	 * @throws HandshakeException if any of the checks fails
 	 */
 	public void verifyCertificate(CertificateMessage message) throws HandshakeException {
-		if (message.getCertificateChain() != null) {
-			if (isClient && message.getCertificateChain().getCertificates().isEmpty()) {
-				LOGGER.debug("Certificate validation failed: empty server certificate!");
-				AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.BAD_CERTIFICATE,
-						session.getPeer());
-				throw new HandshakeException("Empty server certificate!", alert);
-			}
-			if (certificateVerifier != null) {
-				certificateVerifier.verifyCertificate(message, session);
-			} else {
+		CertPath certPath = message.getCertificateChain();
+		if (certPath != null) {
+			if (certificateVerifier == null) {
 				LOGGER.debug("Certificate validation failed: x509 could not be trusted!");
 				AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.UNEXPECTED_MESSAGE,
 						session.getPeer());
 				throw new HandshakeException("Trust is not possible!", alert);
 			}
+
+			List<? extends Certificate> certificates = certPath.getCertificates();
+			if (certificates.isEmpty()) {
+				if (isClient) {
+					LOGGER.debug("Certificate validation failed: empty server certificate!");
+					AlertMessage alert = new AlertMessage(AlertLevel.FATAL, AlertDescription.BAD_CERTIFICATE,
+							session.getPeer());
+					throw new HandshakeException("Empty server certificate!", alert);
+				}
+			}
```

Comment:
This new empty check should not be moved in AdvancedCertificateVerifier ?

Reply:
If your server certificate is empty, I would guess, your handshake will then fail anyway later. Otherwise scandium must explicitly support anonymous server.